src/test/scala/examples/DecoupledAdvTester.scala

```diff
@@ -0,0 +1,69 @@
+// See LICENSE for license details.
+
+package examples
+
+import chisel3._
+import chisel3.util.{EnqIO, DeqIO, Queue}
+import chisel3.iotesters.{ChiselFlatSpec, AdvTester}
+
+class SmallOdds5(filter_width: Int) extends Module {
+
+  class FilterIO extends Bundle {
+    val in = DeqIO(UInt(width = filter_width))
+    val out = EnqIO(UInt(width = filter_width))
+  }
+
+  class Filter(isOk: UInt => Bool) extends Module {
+    val io = IO(new FilterIO)
+
+    io.in.ready := io.out.ready
+    io.out.bits := io.in.bits
+
+    io.out.valid := io.out.ready && io.in.valid && isOk(io.in.bits)
```

Comment:
Ready is supposed to indicate that a sink is ready to accept data, and Valid is supposed to indicate that the data bits are meaningful (valid). The two are not necessarily dependent on each other (and best practices specify that they should not be, I think). A data transfer occurs when both ready and valid are high simultaneously, so any coupling between ready/valid signals would be gated by a register.